[TNT-113] 트레이니 회원 정보 입력 화면 구현

SeonJeongk opened the pull request:
## 📝 작업 내용

- Closes #16 
<br>

- 트레이니 회원가입 상단 공통 부분을 `StepProgressHeader` 컴포넌트로 분리
- 트레이니 이름 입력 화면 UI 구현
- 트레이니 기본 정보 입력 UI 구현
- 트레이니 PT 목적 선택 UI 구현
- 트레이니 PT 주의사항 입력 UI 구현
- 트레이니 프로필 생성 완료 UI 구현
<br>

- TextField 컴포넌트에 키보드 타입, counter (`0/0자`) 보여지기 여부도 설정 가능하도록 수정했습니다
- 유저 프로필 사진 설정 영역 `ProfileImageSection` 컴포넌트로 분리했습니다
- PT 목적 선택 화면은 하단의 `다음` 버튼 클릭 시, 임시로 선택된 버튼들을 Log로 출력하도록 구현했습니다. 테스트용이기 때문에 추후 삭제될 예정입니다.

## 📸 실행 화면
### 트레이니 이름 입력 화면
https://github.com/user-attachments/assets/416322d6-7678-4d62-9a49-74e9b52e205d

### 트레이니 기본 정보 입력 화면
https://github.com/user-attachments/assets/2b6f2c3a-03ee-4fca-91c0-05fca87cb19d

### 트레이니 PT 목적 선택 화면
https://github.com/user-attachments/assets/887515da-ad45-4db6-b8b8-e790104e999c

### 트레이니 PT 주의사항 입력 화면
https://github.com/user-attachments/assets/b2179b8d-5677-41bb-ac17-fabfbb90db9d

### 트레이니 프로필 생성 완료 화면

https://github.com/user-attachments/assets/19608de9-a256-49aa-8681-8fcdaf8b07b1




## 🙆🏻 리뷰 요청 사항
잘 부탁드립니다..


## 👀 레퍼런스
https://developer.android.com/develop/ui/compose/layouts/flow?hl=ko#item-weights

SeonJeongk committed:
[TNT-113] feat: 프로필 이미지 선택 컴포저블 컴포넌트로 분리

SeonJeongk committed:
[TNT-113] feat: 트레이니 회원가입 StepProgress 컴포넌트 구현

- 현재 절차에 해당하는 숫자 indicator 색상 다르게 보여주기
- subTitle은 이름 입력 화면에서는 없기 때문에 nullable하게 처리

SeonJeongk committed:
[TNT-113] feat: 트레이니 이름 입력 UI 구현

SeonJeongk committed:
[TNT-113] refactor: 매개변수 순서 변경

SeonJeongk committed:
[TNT-113] refactor: 트레이니 프로필 설정 화면 파일, 변수명 수정

SeonJeongk committed:
[TNT-113] design: TextField 키보드 타입과 counter 보여지기 여부 추가

SeonJeongk committed:
[TNT-113] feat: 트레이니 이름 입력 UI Scafford로 수정

SeonJeongk committed:
[TNT-113] feat: 트레이니 기본정보 입력 UI 구현

SeonJeongk committed:
[TNT-113] feat: 사용할 string 리소스 추가

SeonJeongk committed:
[TNT-113] feat: 트레이니 PT 주의사항 입력 UI 구현

SeonJeongk committed:
[TNT-113] feat: 트레이니 프로필 생성 완료 UI 구현

SeonJeongk committed:
[TNT-113] feat: 트레이니 PT 목적 화면 UI 구현

SeonJeongk committed:
[TNT-113] fix: 매개변수 순서 변경

SeonJeongk committed:
[TNT-113] fix: 기본 정보 입력 화면 typo 적용

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -0,0 +1,234 @@
+        )
+    }
+}
+
+@Composable
+private fun Unit(stringResId: Int) {
+    Text(
+        text = stringResource(stringResId),
+        style = TnTTheme.typography.body1Medium,
+        color = TnTTheme.colors.neutralColors.Neutral400,
+    )
+}
```
오잉 얘는 어째서 이름이 `Unit` 인가요?

SeonJeongk replied:
kg, cm와 같은 단위라 `Unit`으로 했습니다

hoyahozz replied:
아시겠지만 코틀린에는 이미 `Unit` 이라는 원시타입이 있어서 다른 이름을 생각해보는게 좋을 것 같아요!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -0,0 +1,234 @@
+            .fillMaxWidth()
+            .padding(8.dp)
+            .clickable {
+                val calendar = Calendar.getInstance()
```
`LocalDate` 로 변경해주세요!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -0,0 +1,234 @@
+    var selectedDate by remember { mutableStateOf(initialDate) }
+    val context = LocalContext.current
+
+    val textColor = if (selectedDate == initialDate) {
```
컴포저블 내에서 한 번만 사용되는 케이스라, `Text` 에 바로 주입해줘도 될 것 같아요!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -0,0 +1,234 @@
+@Composable
+private fun BirthdayPicker(
+    modifier: Modifier = Modifier,
+    initialDate: String = "2000/01/01",
```
`LocalDate` 를 주입받고 내부에서 `String` 으로 변환하는 형태가 나올 것 같아요 !

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -0,0 +1,234 @@
+            showCounter = false,
+            keyboardType = KeyboardType.Number,
+            trailingComponent = {
+                Unit(R.string.height_unit)
```
사실 두 번만 반복되고 있는 컴포넌트라서 분리가 필요한지 잘 모르겠어요 ㅎ.ㅎ

그리고 코드를 작성할 때는 글을 쓸 때처럼 위에서 아래로 읽을 수 있도록 작성하는게 좋아요

현재 `Unit` 은 위에 선언되어 있는데 사용하는 곳은 아래쪽이라서 읽는 입장에서 위로 갔다가 아래로 가게 되네요!

SeonJeongk replied:
아래로 읽을 수 있도록 수정하겠습니다
디자인의 프레임 단위로 컴포넌트를 분리하면 나중에 코드를 수정해야 할 일이 생겼을 때 보기 편할거라 생각했습니다! 반복이 많이 없는 부분이긴 합니다..! 위로 합쳐둘게요👍

hoyahozz commented on core/designsystem/src/main/res/values/strings.xml:
```diff
@@ -10,6 +11,12 @@
     <string name="trainee">트레이니</string>
 
+    <string name="user_birthday">생년월일</string>
+    <string name="user_height">키</string>
```
'키' 가 `user_height` 인가용?

SeonJeongk replied:
움.. `(용어)_label`로 통일하겠습니다! `height_label`로 수정해둘게요

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeNoteForTrainerScreen.kt:
```diff
@@ -0,0 +1,82 @@
+                TnTOutlinedTextField(
+                    value = text,
+                    onValueChange = { newValue ->
+                        if (newValue.length <= 100) {
```
이런 곳에서 상수를 선언해주는게 좋습니다!

왜 100보다 적어야하지? 하는 의문이 생길 수 있을 것 같아요

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -0,0 +1,133 @@
+@Composable
+fun TraineePTPurposeScreen() {
+    // Pair로 관리
+    val purposeItems = listOf(
```
이거 `enum` 으로 빼는게 좋아보입니다!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -0,0 +1,133 @@
+    // TODO 리소스 id값 텍스트로 전환해 넘겨주기
+    val selectedPurposes = remember { mutableStateOf(setOf<Int>()) }
+
+    // 선택된 값 확인용
```
이 주석은 없어도 되지 않을까요?

SeonJeongk replied:
넵 지웠습니다

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -0,0 +1,133 @@
+                    verticalArrangement = Arrangement.spacedBy(12.dp),
+                    modifier = Modifier.padding(horizontal = 20.dp),
+                ) {
+                    // 한 행에 2개씩
```
한 행에 두 개를 표현하기 위해 `Pair` 를 사용하신게 맞으실까요?

음.. 이것보다는 좀 더 나은 방법이 있을 것 같아요 !_!

SeonJeongk replied:
`FlowRow`로 수정했습니다!

hoyahozz reviewed (changes requested):
이따 집가서 한 번 더 보겠습니당!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -0,0 +1,133 @@
+
+@Composable
+fun PurposeButton(
+    @StringRes textResId: Int,
```
리소스 대신 텍스트 그대로 받아도 될 것 같습니다!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -0,0 +1,133 @@
+}
+
+// 선택된 값 업데이트
+private fun toggleSelection(selectedPurposes: MutableState<Set<Int>>, purposeId: Int) {
```
`MutableCollection` 을 넘겨주면 어떤 문제가 발생할 수 있을까요?

hoyahozz replied:
'토글' 이라는 간단한 동작인데, 주석이 추가되어야만 했던 이유를 생각해보면 좋을 것 같아요!

좀 더 명확한 로직을 찾아봅시다 !_!

SeonJeongk replied:
`MutableCollection`을 직접 함수에 전달하면 데이터 변경 추적이 어려워지고, 상태의 불변성을 위반할 위험이 있다고 하네요..! 😢

`Set`으로 수정했습니다

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/component/StepProgressHeader.kt:
```diff
@@ -0,0 +1,105 @@
+import co.kr.tnt.designsystem.theme.TnTTheme
+
+@Composable
+fun StepProgressHeader(
```
이 컴포넌트 입장에서 본인이 '헤더' 라는 사실을 인지해야 할까요?

'헤더' 라는 이름을 부여함으로써, 이를 사용하는 모든 곳에서는 해당 컴포넌트를 헤더로만 사용하는 것이 강제될 수 있어요!

이전에도 이야기했듯이, 범용적으로 컴포저블 이름을 구성해봅시돠!

SeonJeongk replied:
움 회원가입에만 쓰일 것 같아서 저렇게 지었습니다!
범용적인 사용을 고려해  `ProgressSteps`로 수정하겠습니다!`

hoyahozz commented on core/designsystem/src/main/java/co/kr/tnt/designsystem/component/TextField.kt:
```diff
@@ -119,8 +125,10 @@ fun TnTLabeledTextField(
     maxLength: Int = 15,
     isSingleLine: Boolean = false,
     showWarning: Boolean = false,
+    showCounter: Boolean = true,
```
흐음 이쯤되니까 `counter` 있는 컴포넌트와 없는 컴포넌트를 분리하는게 나을 것 같다는 생각이 들긴 하네요 '-'

요거 나중에 분리해버립시다!

SeonJeongk committed:
[TNT-113] fix: StepProgressHeader 이름 수정

SeonJeongk committed:
[TNT-113] fix: 날짜 LocalDate로 수정

SeonJeongk committed:
[TNT-113] fix: 주의사항 입력 제한 상수로 선언

SeonJeongk committed:
[TNT-113] fix: 키, 몸무게 입력 제한 상수로 선언

SeonJeongk committed:
[TNT-113] fix: string 리소스 명 수정

SeonJeongk committed:
[TNT-113] refactor: PT 목적 선택 화면 리뷰 반영

- enum으로 목적 관리
- 버튼 FlowRow로 보여주기
- 버튼에 텍스트 그대로 넘겨주기
- toggleSelection Set으로 넘겨주도록 수정
- toggleSelection 로직 수정

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -28,22 +27,16 @@ import co.kr.tnt.designsystem.component.button.model.ButtonSize
-    // 선택된 값 확인용
-    val context = LocalContext.current
-    val selectedTexts = selectedPurposes.value.map { context.getString(it) }
+    val selectedPurposes = remember { mutableStateOf(setOf<PTPurpose>()) }
```
`by remember` 와 같은 위임 형태로 선언하면, `.value` 를 사용해야 하는 번거로움을 줄일 수 있어요 ㅎㅎ

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -59,26 +52,20 @@ fun TraineePTPurposeScreen() {
-                                modifier = Modifier.weight(1f),
-                            )
+                Column(modifier = Modifier.padding(horizontal = 20.dp)) {
+                    FlowRow(
```
이쪽 코드 너무 깔끔해진 것 같아요 👍 👍👍

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -114,59 +149,50 @@ fun TraineeBasicInfoScreen() {
-                val year = calendar.get(Calendar.YEAR)
-                val month = calendar.get(Calendar.MONTH)
-                val day = calendar.get(Calendar.DAY_OF_MONTH)
+                val today = LocalDate.now()
```
`click` 할 때마다 `LocalDate` 가 생성될 것 같아요! 최대한 상단에서 선언해주는게 좋을 것 같습니당

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -114,59 +149,50 @@ fun TraineeBasicInfoScreen() {
                 ).apply {
                     // 오늘 이후는 선택 불가능
-                    datePicker.maxDate = today.timeInMillis
+                    datePicker.maxDate = Calendar.getInstance().timeInMillis
```
`LocalDate` 로 밀리세컨드 추출 가능합니다 !_!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineeBasicInfoScreen.kt:
```diff
@@ -114,59 +149,50 @@ fun TraineeBasicInfoScreen() {
-                    year,
-                    month,
-                    day,
+                    today.year,
+                    today.monthValue - 1,
+                    today.dayOfMonth,
```
선택한 날짜가 있음에도 다시 다이얼로그로 진입하면 '오늘' 날짜로 세팅될 것 같아요!

선택한 날짜가 있는 경우 다이얼로그로 진입하면 그 날짜가 기본 세팅값이 되도록 수정하면 좋을 것 같아요!

SeonJeongk replied:
앗 그러네요 `selectedDate` 이용해 null이면 오늘 날짜로, 값이 있으면 해당 날짜가 기본 세팅 값이 되도록 수정했습니다!

hoyahozz reviewed (approved):
고생하셨습니다! 👍👍👍

추가로 남긴 코멘트 확인 및 수정만 해주시고, 정상 작동하면 바로 머지 부탁드립니다~~!

hoyahozz commented on feature/signup/src/main/java/co/kr/tnt/signup/trainee/TraineePTPurposeScreen.kt:
```diff
@@ -0,0 +1,119 @@
+    purpose: PTPurpose,
+): Set<PTPurpose> {
+    return selectedPurposes.toMutableSet().apply {
+        if (contains(purpose)) remove(purpose) else add(purpose)
```
`if`, `else` 구문 사용 시 줄바꿈하면 좋을 것 같아요 ~

SeonJeongk committed:
[TNT-113] fix: PT 목적 화면 리뷰 반영

- by remember로 변경
- if-else문 줄바꿈

SeonJeongk committed:
[TNT-113] fix: 트레이니 기본 정보 입력 화면 리뷰 반영

- LocalDate 상단에 선언하기 (today)
- 선택된 날짜가 있으면 해당 날짜를 기본 세팅 값으로 설정
- LocalDate로 밀리세컨드 추출
- if-else문 줄바꿈
